src/backend/controllers/marketController.js

```diff
@@ -0,0 +1,308 @@
+// @flow strict
+
+// Package Imports
+import type {
+    $Request,
+    $Response,
+    NextFunction,
+    Middleware
+} from 'express';
+import { validationResult } from 'express-validator';
+
+// MarketplaceSale Model, User Model, Tank Model
+import MarketSale from '../../models/marketSaleModel';
+import User from '../../models/userModel';
+import Tank from '../../models/tankModel';
+
+// Adds a Marketplace Sale
+// Different branches depending the itemType
+exports.addMarketSale = async (req: $Request, res: $Response) => {
+    // Contains errors for failed validation.
+    const errors = validationResult(req);
+
+    if (!errors.isEmpty()) {
+        // Return 400 for a bad request
+        return res
+            .status(400)
+            .json({ errors: errors.array() });
+    }
+    
+    // Deconstructing information from request body
+    // Note: Depending on the itemType, if it is a tank,
+    // itemID is the ObjectId of the tank, otherwise it is
+    // the string identifier of the component/Casus block equivalent
+    // to the property in the User Model.
+    const { sellerId, salePrice, itemId, itemType, amount } = req.body;
+
+    // If the item is a tank
+    if (itemType === 'tank') {
+        try { 
+            // Check if tank is actually owned by user
+            // Also acts as a way to check if user is real
+            let tank = await Tank.findOne({ id: itemId, userId: sellerId });
+            if (!tank) {
+                return res
+                    .status(400)
+                    .json({ msg: 'Tank does not exist under this user.' })
+            }            
+
+            // Make a new Marketplace Sale
+            // if tank exists
+            const sale = new MarketSale({
+                sellerId,
+                salePrice,
+                itemId,
+                itemType,
+                amount 
+            });
+
+            // Mark tank as being sold and save change to DB
+            tank.userId = 'Marketplace';
+            await tank.save();
+
+            // Save the sale to DB
+            await sale.save();
+
+            // Send back success confirmation
+            res.status(201).json({ msg: 'Successfully created Market Sale.' })           
+
+        } catch (err) {
+            console.error(err.message);
+            res.status(500).json({ msg: 'Unable to add Market Sale.' });
+        }
+    } else { // If the item is a component or casus block
+        try {
+            // Check if user exists
+            let user = await User.findById(sellerId);
+            if (!user) {
+                return res
+                    .status(400)
+                    .json({ msg: 'User not found in DB' })
+            }
+
+            if (itemType === 'component') { // Components
+                // Check if they have enough
+                // of the component to sell
+                const userItem = user['inventory']['tankComponents'][itemId];
```

Comment:
You don't verify that itemId is a legal string, do you?

Is that a problem?

Reply:
I can, just need to add something to the checks in the route.